conda/__init__.py

```diff
@@ -148,3 +139,209 @@ def _default(self, obj):
+        self._argument = argument
+        self._rename = rename
+
+        return self
+
+    @staticmethod
+    def _get_module(stack: int) -> tuple[ModuleType, str]:
+        import inspect  # expensive
+
+        try:
+            frame = inspect.stack()[2 + stack]
+            module = inspect.getmodule(frame[0])
+            return (module, module.__name__)
+        except (IndexError, AttributeError):
+            raise RuntimeError("unable to determine the calling module") from None
+
+    @classmethod
+    def module(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for modules."""
+        cls.topic(
+            deprecate_in=deprecate_in,
+            remove_in=remove_in,
+            topic=cls._get_module(stack)[1],
+            addendum=addendum,
+            stack=stack + 2,
+        )
+
+    @classmethod
+    def constant(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        constant: str,
+        value: Any,
+        *,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for module constant (global)."""
+        # detect calling module
+        module, fullname = cls._get_module(stack)
+
+        self = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(f"{fullname}.{constant}"))
+
+        # patch module level __getattr__ to alert user that it's time to remove something
+        super_getattr = getattr(module, "__getattr__", None)
+
+        def __getattr__(name: str) -> Any:
+            if name == constant:
+                warnings.warn(
+                    self._message(f"{fullname}.{name}"), self._category, stacklevel=2 + self._stack
+                )
+                return value
+
+            if super_getattr:
+                return super_getattr(name)
+
+            raise AttributeError(f"module '{fullname}' has no attribute '{name}'")
+
+        module.__getattr__ = __getattr__
+
+    @classmethod
+    def topic(
+        cls,
+        deprecate_in: str,
+        remove_in: str,
+        *,
+        topic: str,
+        addendum: str | None = None,
+        stack: int = 0,
+    ) -> None:
+        """Deprecation function for a topic."""
+        self = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)
+
+        # alert developer that it's time to remove something
+        if not self._category:
+            raise RuntimeError(self._message(topic))
+
+        # alert user that it's time to remove something
+        warnings.warn(self._message(topic), self._category, stacklevel=2 + self._stack)
```

Comment:
```suggestion
        deprecation = cls(deprecate_in=deprecate_in, remove_in=remove_in, addendum=addendum, stack=stack)

        # alert developer that it's time to remove something
        if not deprecation._category:
            raise RuntimeError(deprecation._message(topic))

        # alert user that it's time to remove something
        warnings.warn(deprecation._message(topic), deprecation._category, stacklevel=2 + deprecation._stack)
```

Reply:
It's a little weird to me to see `self` used in a class method

Reply:
gotcha, I like to do this as it makes it very clear (to me) that it's an instance of the `cls` as opposed to being some other object 🤷🏼‍♂️ but I can see how it may be perceived as confusing

Reply:
Yeah, I understand, looking how this could be confusing during debugging etc mostly.